### ISSUE TITLE
proj_context_guess_dialect

### ISSUE THREAD

**Issue by snowman2** [OSGeo/PROJ#1200]
Is there something analogous to `proj_context_guess_wkt_dialect` for determining the `PJ_PROJ_STRING_TYPE`? 

If not, this would be nice to have added to `proj.h`. One potential name could be `proj_context_guess_dialect`.

**Comment by rouault** [comment]
No, there isn't such capability.

**Comment by rouault** [comment]
I don't have such need, so unlikely to come from me. Why would that be needed ?
Anyway, from my experience, I would discourage use of PROJ strings at all to express CRS objects for example because they aren't rich enough for that, and suffer from ambiguities (axis order, units, etc,) I tried to remedy a bit with the "PROJ 5" variations, but that's far from being perfect. PROJ strings however are much better used to express coordinate transformation / pipelines.

**Comment by snowman2** [comment]
The idea would be to modify a parameter in the string based on the version. For example, in `pyproj` it has a `preserve_units` option where it makes sure the `+units` parameter is `+units=m` (I don't know the history or reasoning behind it, it would be for backwards compatibility). However, that syntax is non-existent in the version 5 of the proj string and a different set of parameters would be needed to be modified.

>> I would discourage use of PROJ strings at all to express CRS objects

That is good to know. However, for compatibility with libraries that do so such as (rasterio, fiona, geopandas, etc...) it is a useful feature to have. This would be useful to have documented somewhere.

**Comment by kbevers** [comment]
> However, that syntax is non-existent in the version 5 of the proj string

What syntax is that exactly?

> However, for compatibility with libraries that do so such as (rasterio, fiona, geopandas, etc...) it is a useful feature to have

That's why it's still there ;) It is likely though that a future version of PROJ removes support for parameters such as `+datum` and `+towgs84`.

**Comment by snowman2** [comment]
The original pyproj version adjusts it [here](https://github.com/jswhit/pyproj/blob/78540f5ff40da92160f80860416c91ee74b7643c/lib/pyproj/__init__.py#L352).

The workaround I have is [here](https://github.com/snowman2/pyproj/blob/3663f6dcc83746ccd948a215799240aa7b16648e/pyproj/__init__.py#L401).

**Comment by kbevers** [comment]
You didn't answer my question? What syntax is it that you think is unavailable in PROJ 5 compared to PROJ.4? `+units`? That's supported in PROJ 5.  `preserve_units` is not a PROJ construct.

**Comment by snowman2** [comment]
Hi @kbevers, thanks for following up. I was thinking the links I provided gave more insight into the matter.
I think I am not being super clear about what behavior I am attempting to accomplish (mostly because I am not familiar enough with the `preserve_units` argument in the `pyproj` world).

In the examples in the `pyproj.Proj` class it has an example:
```
>>> from pyproj import Proj
>>> p1 = Proj(init="epsg:32667")
>>> 'x=%12.3f y=%12.3f (meters)' % p1(-114.057222, 51.045)
'x=-1783486.760 y= 6193833.196 (meters)'
>>> p2 = Proj("+init=epsg:32667",preserve_units=True)
>>> 'x=%12.3f y=%12.3f (feet)' % p2(-114.057222, 51.045)
'x=-5851322.810 y=20320934.409 (feet)'
```
When looking at the original projection:
```
>>> from pyproj import CRS
>>> crs = CRS("+init=epsg:32667")
>>> crs.to_string(5)
'+proj=pipeline +step +proj=axisswap +order=2,1 +step +proj=unitconvert +xy_in=deg +xy_out=rad +step +proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +ellps=WGS84 +step +proj=unitconvert +xy_in=m +z_in=m +xy_out=us-ft +z_out=us-ft'
>>> crs.to_string(4)
'+proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +datum=WGS84 +units=us-ft +no_defs'
```
If I try to add `+units=m` to the version 5 string, I get errors when attempting to create a projection from it. It appears that the `+xy_in=m +z_in=m +xy_out=us-ft +z_out=us-ft` is the new replacement for `+units`.
So, my band-aid solution was to convert to a version 4 string and modify the `+units` parameter there and re-create a projection with the new string.
```
>>> new_crs = '+proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +datum=WGS84 +units=m +no_defs'
>>> crs2 = CRS(new_crs)
>>> crs2.to_string(5)
'+proj=pipeline +step +proj=unitconvert +xy_in=deg +xy_out=rad +step +proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +ellps=WGS84'
>>> crs2.to_string(4)
'+proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +datum=WGS84 +units=m +no_defs'
```
The conversion seems to work, but the original (non-converted example) does not seem to work. 
```
>>> from pyproj import Proj
>>> p1 = Proj(init="epsg:32667")
>>> 'x=%12.3f y=%12.3f (meters)' % p1(-114.057222, 51.045)
'x=-1783486.760 y= 6193833.196 (meters)'
>>> p2 = Proj("+init=epsg:32667",preserve_units=True)
>>> 'x=%12.3f y=%12.3f (feet)' % p2(-114.057222, 51.045)
'x=1000000000000000019884624838656.000 y=1000000000000000019884624838656.000 (feet)'
```
I am not a proj.4 expert. So, if you have suggestions, I would be happy to hear them. Thanks!

**Comment by snowman2** [comment]
`proj` seems to handle it fine, so it may be how `pyproj` is handling it.
```
$ pyproj/proj_dir/bin/proj "+init=epsg:32667"
-114.057222 51.045     
-5851322.81	20320934.41
```

**Comment by kbevers** [comment]
Okay, now I understand. I assmue you use `proj_create_crs_to_crs()` for this. I don't think it make much sense to try to convert coordinate units into meters when using EPSG CRS definitions. PROJ returns coordinates in the correct units according to the specification of that particular system. If you somehow still think this is a good idea you could add another step to the pipeline:

`+proj=pipeline +step +proj=axisswap +order=2,1 +step +proj=unitconvert +xy_in=deg +xy_out=rad +step +proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +ellps=WGS84 +step +proj=unitconvert +xy_in=m +z_in=m +xy_out=us-ft +z_out=us-ft`
-> 
`+proj=pipeline +step +proj=axisswap +order=2,1 +step +proj=unitconvert +xy_in=deg +xy_out=rad +step +proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +ellps=WGS84 +step +proj=unitconvert +xy_in=m +z_in=m +xy_out=us-ft +z_out=us-ft +step +proj=unitconvert +xy_in=us-ft +xy_out=m`

or you could simply just remove the last step of the pipeline:

`+proj=pipeline +step +proj=axisswap +order=2,1 +step +proj=unitconvert +xy_in=deg +xy_out=rad +step +proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +ellps=WGS84`

I wouldn't bother with the above though, I'd just accept what comes out of PROJ as it is and also create a pyproj function that wraps `proj_create()` so that power users can make their own transformations having full control over what happens. Overriding the output units of a predefined CRS is a good use case for that.

**Comment by snowman2** [comment]
@kbevers thanks for your response. I am not entirely sure why the `preserve_units` exists, but I think I would need to understand more about why it is there before I remove it. The good news is that the `pyproj.Proj` class should support power users with the preserve_units=True set.

**Comment by snowman2** [comment]
I have been doing some testing on the example that inspired this issue and have been doing some digging. And here is what I have found so far.

So, if I get the proj 5 string from projinfo:
```
$ ./pyproj/proj_dir/bin/projinfo epsg:32667
PROJ string:
+proj=pipeline +step +proj=axisswap +order=2,1 +step +proj=unitconvert +xy_in=deg +xy_out=rad +step +proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +ellps=WGS84 +step +proj=unitconvert +xy_in=m +z_in=m +xy_out=us-ft +z_out=us-ft
...
```
And try to use the `proj` command, I get this error:
```
$ ./pyproj/proj_dir/bin/proj +proj=pipeline +step +proj=axisswap +order=2,1 +step +proj=unitconvert +xy_in=deg +xy_out=rad +step +proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996 +x_0=500000.001016002 +y_0=0 +ellps=WGS84 +step +proj=unitconvert +xy_in=m +z_in=m +xy_out=us-ft +z_out=us-ft
Rel. 6.0.0, March 1st, 2019
<proj>: 
can't initialize operations that take angular input coordinates
program abnormally terminated
```
But somehow this works:
```
$ ./pyproj/proj_dir/bin/proj +init=epsg:32667
-114.057222 51.045 
-5851322.81	20320934.41 
```
Is this behavior expected?

**Comment by kbevers** [comment]
No, there are several issues at play. First, the error message is providing wrong information. Fixed in fbc8d77.

As for the `+init` initialization, it works because the string is created in "PROJ.4 mode":

https://github.com/OSGeo/proj.4/blob/fbc8d776d75deccf98296bc6406e5be68dcef3f2/src/pj_init.c#L288-L291

which you can also see when calling `proj` in verbose mode:

```
$ ./bin/proj -V +init=epsg:32667
#Transverse Mercator
#	Cyl, Sph&Ell
# +init=epsg:32667 +proj=tmerc +lat_0=0 +lon_0=-81 +k=0.9996
# +x_0=500000.001016002 +y_0=0 +datum=WGS84 +units=us-ft +no_defs
# +ellps=WGS84 +towgs84=0,0,0
#Final Earth figure: ellipsoid
#  Major axis (a): 6378137.000
#  1/flattening: 298.257224
#  squared eccentricity: 0.006694379990
```

Using the PROJ 5 pipeline fails, rightly so, because the first step in the pipeline takes non-angular input but `proj` expects angular input units. This can be fixed by adding `+angular` to the `axisswap` pipeline step. 

@rouault Do you reckon this is an easy fix? If so I'd like use this is a hands-on exercise in getting more familiar with the new code :)

**Comment by rouault** [comment]
> This can be fixed by adding `+angular` to the `axisswap` pipeline step.

Adding +angular seems to be a repetition since +xy_in=deg +xy_out=rad clearly indicates that those are angular units, no ?

> Do you reckon this is an easy fix?

I think the fix would be in the axisswap operation itself, no ?

**Comment by kbevers** [comment]
> Adding +angular seems to be a repetition since +xy_in=deg +xy_out=rad clearly indicates that those are angular units, no ?

That's for the `unitconvert` operation which is the second step in the pipeline. It doesn't affect the first step of `axisswap`, which is why the `+angularunits` is needed. 

> I think the fix would be in the axisswap operation itself, no ?

Not really, no. I would assume this should be handled somewhere in the database logic. It *might* also be possible to handle in the pipeline code.

**Comment by kbevers** [comment]
I have a fix in #1205 that at least works for the specific case mentioned here. I *think* that it is generic enough that it handles all but the most ridiculous pipelines.
